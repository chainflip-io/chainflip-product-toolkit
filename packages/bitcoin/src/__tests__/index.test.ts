import { describe, it, expect } from 'vitest';
import { encodeAddress, isValidAddressForNetwork } from '../index';

const networks = ['mainnet', 'perseverance', 'sisyphos', 'backspin', 'regtest', 'testnet'] as const;

const cases = (
  [
    ['P2WPKH', '0x63b7cc4432bba5c51fd09428c47abaf2d52f9373'],
    ['P2SH', '0x86bc2bfa0fdb414245561008c08615f4632604f9'],
    ['P2PKH', '0xecd30fe4f116f7485d7fd2f1cf26d0a059561548'],
    ['P2WSH', '0x321e6259700f0734a61b5ffa849de17865bc3293d5d069d91f23e6cb87daedf0'],
    ['Taproot', '0x68a3db628eea903d159131fcb4a1f6ed0be6980c4ff42b80d5229ea26a38439e'],
  ] as const
).flatMap(([type, address]) => networks.map((n) => [type, n, address] as const));

describe('bitcoin', () => {
  it.each(cases)('decodes a %s address on %s', (kind, network, address) => {
    expect(encodeAddress(address, kind, network)).toMatchSnapshot(`${kind} on ${network}`);
  });

  it('throws for invalid networks', () => {
    expect(() =>
      encodeAddress('0x63b7cc4432bba5c51fd09428c47abaf2d52f9373', 'P2WPKH', 'invalid' as any),
    ).toThrowErrorMatchingInlineSnapshot(`[Error: Invalid network: invalid]`);
  });

  it('throws for invalid address types', () => {
    expect(() =>
      encodeAddress('0x63b7cc4432bba5c51fd09428c47abaf2d52f9373', 'invalid' as any, 'mainnet'),
    ).toThrowErrorMatchingInlineSnapshot(`[Error: Invalid address type: invalid]`);
  });

  it('throws for invalid addresses', () => {
    expect(() =>
      encodeAddress('0x63b7cc4432bba5c51fd09428c47abaf2d52f937', 'P2WPKH', 'mainnet'),
    ).toThrowErrorMatchingInlineSnapshot(`[Error: bytes must have an even number of characters]`);
  });

  it('throws for non-hex addresses', () => {
    expect(() =>
      encodeAddress('0x63b7cc4432bba5c51fd09428c47abaf2d52f937g', 'P2WPKH', 'mainnet'),
    ).toThrowErrorMatchingInlineSnapshot(`[Error: bytes must be hex-encoded with a 0x prefix]`);
  });

  it('throws on bad addresses', () => {
    expect(() => encodeAddress('0x00', 'Taproot', 'mainnet')).toThrowErrorMatchingInlineSnapshot(
      `[Error: Invalid address]`,
    );
  });
});

describe(isValidAddressForNetwork, () => {
  const bitcoinAddresses = {
    mainnet: {
      P2PKH: ['1PYVSoeftFP4EVBN3ou8vZctkhDthJamvp'],
      P2SH: ['32k55FA93MYqbjhLk9hokD3P666Vz9QqKb'],
      Taproot: ['bc1qwqdg6squsna38e46795at95yu9atm8azzmyvckulcc7kytlcckxswvvzej'],
    },
    testnet: {
      P2PKH: ['mipcBbFg9gMiCh81Kj8tqqdgoZub1ZJRfn', 'n2ZtW11kqP6pf3umiU4NaSRgrL9aQJVuJV'],
      P2SH: ['2MzQwSSnBHWHqSAqtTVQ6v47XtaisrJa1Vc'],
      Taproot: ['tb1qw508d6qejxtdg4y5r3zarvary0c5xw7kxpjzsx'],
    },
    regtest: {
      Taproot: ['bcrt1qs758ursh4q9z627kt3pp5yysm78ddny6txaqgw'],
    },
  };

  describe.each(Object.entries(bitcoinAddresses))('for %s', (network, addressTypes) => {
    describe.each(Object.entries(addressTypes))('for %s', (kind, addresses) => {
      it.each(addresses)('gets the network for %s', (address) => {
        expect(
          isValidAddressForNetwork(address, network as 'mainnet' | 'testnet' | 'regtest'),
        ).toBe(true);
      });
    });
  });

  it('returns null for invalid addresses', () => {
    expect(isValidAddressForNetwork('invalid', 'mainnet')).toBe(false);
  });
});
