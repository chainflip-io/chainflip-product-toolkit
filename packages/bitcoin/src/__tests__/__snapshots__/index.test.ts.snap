// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`bitcoin > decodes a P2PKH address on backspin > P2PKH on backspin 1`] = `"n37AcBDN48pKxBR5DK1fSDFnzFuMhGq9wy"`;

exports[`bitcoin > decodes a P2PKH address on backspin > P2PKH on backspin 2`] = `"n37AcBDN48pKxBR5DK1fSDFnzFuMhGq9wy"`;

exports[`bitcoin > decodes a P2PKH address on backspin > P2PKH on backspin 3`] = `"n37AcBDN48pKxBR5DK1fSDFnzFuMhGq9wy"`;

exports[`bitcoin > decodes a P2PKH address on mainnet > P2PKH on mainnet 1`] = `"1NbDK88PF7P5B4wTVk3HcJ3U8GJeiSrccq"`;

exports[`bitcoin > decodes a P2PKH address on mainnet > P2PKH on mainnet 2`] = `"1NbDK88PF7P5B4wTVk3HcJ3U8GJeiSrccq"`;

exports[`bitcoin > decodes a P2PKH address on mainnet > P2PKH on mainnet 3`] = `"1NbDK88PF7P5B4wTVk3HcJ3U8GJeiSrccq"`;

exports[`bitcoin > decodes a P2PKH address on perseverance > P2PKH on perseverance 1`] = `"n37AcBDN48pKxBR5DK1fSDFnzFuMhGq9wy"`;

exports[`bitcoin > decodes a P2PKH address on perseverance > P2PKH on perseverance 2`] = `"n37AcBDN48pKxBR5DK1fSDFnzFuMhGq9wy"`;

exports[`bitcoin > decodes a P2PKH address on perseverance > P2PKH on perseverance 3`] = `"n37AcBDN48pKxBR5DK1fSDFnzFuMhGq9wy"`;

exports[`bitcoin > decodes a P2PKH address on regtest > P2PKH on regtest 1`] = `"n37AcBDN48pKxBR5DK1fSDFnzFuMhGq9wy"`;

exports[`bitcoin > decodes a P2PKH address on regtest > P2PKH on regtest 2`] = `"n37AcBDN48pKxBR5DK1fSDFnzFuMhGq9wy"`;

exports[`bitcoin > decodes a P2PKH address on regtest > P2PKH on regtest 3`] = `"n37AcBDN48pKxBR5DK1fSDFnzFuMhGq9wy"`;

exports[`bitcoin > decodes a P2PKH address on sisyphos > P2PKH on sisyphos 1`] = `"n37AcBDN48pKxBR5DK1fSDFnzFuMhGq9wy"`;

exports[`bitcoin > decodes a P2PKH address on sisyphos > P2PKH on sisyphos 2`] = `"n37AcBDN48pKxBR5DK1fSDFnzFuMhGq9wy"`;

exports[`bitcoin > decodes a P2PKH address on sisyphos > P2PKH on sisyphos 3`] = `"n37AcBDN48pKxBR5DK1fSDFnzFuMhGq9wy"`;

exports[`bitcoin > decodes a P2PKH address on testnet > P2PKH on testnet 1`] = `"n37AcBDN48pKxBR5DK1fSDFnzFuMhGq9wy"`;

exports[`bitcoin > decodes a P2PKH address on testnet > P2PKH on testnet 2`] = `"n37AcBDN48pKxBR5DK1fSDFnzFuMhGq9wy"`;

exports[`bitcoin > decodes a P2PKH address on testnet > P2PKH on testnet 3`] = `"n37AcBDN48pKxBR5DK1fSDFnzFuMhGq9wy"`;

exports[`bitcoin > decodes a P2SH address on backspin > P2SH on backspin 1`] = `"2N5XdxPojGTLjs2ZvkpbF9473zL1wftGzPJ"`;

exports[`bitcoin > decodes a P2SH address on backspin > P2SH on backspin 2`] = `"2N5XdxPojGTLjs2ZvkpbF9473zL1wftGzPJ"`;

exports[`bitcoin > decodes a P2SH address on backspin > P2SH on backspin 3`] = `"2N5XdxPojGTLjs2ZvkpbF9473zL1wftGzPJ"`;

exports[`bitcoin > decodes a P2SH address on mainnet > P2SH on mainnet 1`] = `"3DyRteshezqPfEwP5gyNX77nmyomu5GW2U"`;

exports[`bitcoin > decodes a P2SH address on mainnet > P2SH on mainnet 2`] = `"3DyRteshezqPfEwP5gyNX77nmyomu5GW2U"`;

exports[`bitcoin > decodes a P2SH address on mainnet > P2SH on mainnet 3`] = `"3DyRteshezqPfEwP5gyNX77nmyomu5GW2U"`;

exports[`bitcoin > decodes a P2SH address on perseverance > P2SH on perseverance 1`] = `"2N5XdxPojGTLjs2ZvkpbF9473zL1wftGzPJ"`;

exports[`bitcoin > decodes a P2SH address on perseverance > P2SH on perseverance 2`] = `"2N5XdxPojGTLjs2ZvkpbF9473zL1wftGzPJ"`;

exports[`bitcoin > decodes a P2SH address on perseverance > P2SH on perseverance 3`] = `"2N5XdxPojGTLjs2ZvkpbF9473zL1wftGzPJ"`;

exports[`bitcoin > decodes a P2SH address on regtest > P2SH on regtest 1`] = `"2N5XdxPojGTLjs2ZvkpbF9473zL1wftGzPJ"`;

exports[`bitcoin > decodes a P2SH address on regtest > P2SH on regtest 2`] = `"2N5XdxPojGTLjs2ZvkpbF9473zL1wftGzPJ"`;

exports[`bitcoin > decodes a P2SH address on regtest > P2SH on regtest 3`] = `"2N5XdxPojGTLjs2ZvkpbF9473zL1wftGzPJ"`;

exports[`bitcoin > decodes a P2SH address on sisyphos > P2SH on sisyphos 1`] = `"2N5XdxPojGTLjs2ZvkpbF9473zL1wftGzPJ"`;

exports[`bitcoin > decodes a P2SH address on sisyphos > P2SH on sisyphos 2`] = `"2N5XdxPojGTLjs2ZvkpbF9473zL1wftGzPJ"`;

exports[`bitcoin > decodes a P2SH address on sisyphos > P2SH on sisyphos 3`] = `"2N5XdxPojGTLjs2ZvkpbF9473zL1wftGzPJ"`;

exports[`bitcoin > decodes a P2SH address on testnet > P2SH on testnet 1`] = `"2N5XdxPojGTLjs2ZvkpbF9473zL1wftGzPJ"`;

exports[`bitcoin > decodes a P2SH address on testnet > P2SH on testnet 2`] = `"2N5XdxPojGTLjs2ZvkpbF9473zL1wftGzPJ"`;

exports[`bitcoin > decodes a P2SH address on testnet > P2SH on testnet 3`] = `"2N5XdxPojGTLjs2ZvkpbF9473zL1wftGzPJ"`;

exports[`bitcoin > decodes a P2WPKH address on backspin > P2WPKH on backspin 1`] = `"bcrt1qvwmuc3pjhwju287sjs5vg7467t2jlymnnaxr8m"`;

exports[`bitcoin > decodes a P2WPKH address on backspin > P2WPKH on backspin 2`] = `"bcrt1qvwmuc3pjhwju287sjs5vg7467t2jlymnnaxr8m"`;

exports[`bitcoin > decodes a P2WPKH address on backspin > P2WPKH on backspin 3`] = `"bcrt1qvwmuc3pjhwju287sjs5vg7467t2jlymnnaxr8m"`;

exports[`bitcoin > decodes a P2WPKH address on mainnet > P2WPKH on mainnet 1`] = `"bc1qvwmuc3pjhwju287sjs5vg7467t2jlymnmjyatp"`;

exports[`bitcoin > decodes a P2WPKH address on mainnet > P2WPKH on mainnet 2`] = `"bc1qvwmuc3pjhwju287sjs5vg7467t2jlymnmjyatp"`;

exports[`bitcoin > decodes a P2WPKH address on mainnet > P2WPKH on mainnet 3`] = `"bc1qvwmuc3pjhwju287sjs5vg7467t2jlymnmjyatp"`;

exports[`bitcoin > decodes a P2WPKH address on perseverance > P2WPKH on perseverance 1`] = `"tb1qvwmuc3pjhwju287sjs5vg7467t2jlymn35lwsj"`;

exports[`bitcoin > decodes a P2WPKH address on perseverance > P2WPKH on perseverance 2`] = `"tb1qvwmuc3pjhwju287sjs5vg7467t2jlymn35lwsj"`;

exports[`bitcoin > decodes a P2WPKH address on perseverance > P2WPKH on perseverance 3`] = `"tb1qvwmuc3pjhwju287sjs5vg7467t2jlymn35lwsj"`;

exports[`bitcoin > decodes a P2WPKH address on regtest > P2WPKH on regtest 1`] = `"bcrt1qvwmuc3pjhwju287sjs5vg7467t2jlymnnaxr8m"`;

exports[`bitcoin > decodes a P2WPKH address on regtest > P2WPKH on regtest 2`] = `"bcrt1qvwmuc3pjhwju287sjs5vg7467t2jlymnnaxr8m"`;

exports[`bitcoin > decodes a P2WPKH address on regtest > P2WPKH on regtest 3`] = `"bcrt1qvwmuc3pjhwju287sjs5vg7467t2jlymnnaxr8m"`;

exports[`bitcoin > decodes a P2WPKH address on sisyphos > P2WPKH on sisyphos 1`] = `"tb1qvwmuc3pjhwju287sjs5vg7467t2jlymn35lwsj"`;

exports[`bitcoin > decodes a P2WPKH address on sisyphos > P2WPKH on sisyphos 2`] = `"tb1qvwmuc3pjhwju287sjs5vg7467t2jlymn35lwsj"`;

exports[`bitcoin > decodes a P2WPKH address on sisyphos > P2WPKH on sisyphos 3`] = `"tb1qvwmuc3pjhwju287sjs5vg7467t2jlymn35lwsj"`;

exports[`bitcoin > decodes a P2WPKH address on testnet > P2WPKH on testnet 1`] = `"tb1qvwmuc3pjhwju287sjs5vg7467t2jlymn35lwsj"`;

exports[`bitcoin > decodes a P2WPKH address on testnet > P2WPKH on testnet 2`] = `"tb1qvwmuc3pjhwju287sjs5vg7467t2jlymn35lwsj"`;

exports[`bitcoin > decodes a P2WPKH address on testnet > P2WPKH on testnet 3`] = `"tb1qvwmuc3pjhwju287sjs5vg7467t2jlymn35lwsj"`;

exports[`bitcoin > decodes a P2WSH address on backspin > P2WSH on backspin 1`] = `"bcrt1qxg0xyktspurnffsmtlagf80p0pjmcv5n6hgxnkgly0nvhp76ahcq825u3r"`;

exports[`bitcoin > decodes a P2WSH address on backspin > P2WSH on backspin 2`] = `"bcrt1qxg0xyktspurnffsmtlagf80p0pjmcv5n6hgxnkgly0nvhp76ahcq825u3r"`;

exports[`bitcoin > decodes a P2WSH address on backspin > P2WSH on backspin 3`] = `"bcrt1qxg0xyktspurnffsmtlagf80p0pjmcv5n6hgxnkgly0nvhp76ahcq825u3r"`;

exports[`bitcoin > decodes a P2WSH address on mainnet > P2WSH on mainnet 1`] = `"bc1qxg0xyktspurnffsmtlagf80p0pjmcv5n6hgxnkgly0nvhp76ahcqamg47k"`;

exports[`bitcoin > decodes a P2WSH address on mainnet > P2WSH on mainnet 2`] = `"bc1qxg0xyktspurnffsmtlagf80p0pjmcv5n6hgxnkgly0nvhp76ahcqamg47k"`;

exports[`bitcoin > decodes a P2WSH address on mainnet > P2WSH on mainnet 3`] = `"bc1qxg0xyktspurnffsmtlagf80p0pjmcv5n6hgxnkgly0nvhp76ahcqamg47k"`;

exports[`bitcoin > decodes a P2WSH address on perseverance > P2WSH on perseverance 1`] = `"tb1qxg0xyktspurnffsmtlagf80p0pjmcv5n6hgxnkgly0nvhp76ahcq2n76ye"`;

exports[`bitcoin > decodes a P2WSH address on perseverance > P2WSH on perseverance 2`] = `"tb1qxg0xyktspurnffsmtlagf80p0pjmcv5n6hgxnkgly0nvhp76ahcq2n76ye"`;

exports[`bitcoin > decodes a P2WSH address on perseverance > P2WSH on perseverance 3`] = `"tb1qxg0xyktspurnffsmtlagf80p0pjmcv5n6hgxnkgly0nvhp76ahcq2n76ye"`;

exports[`bitcoin > decodes a P2WSH address on regtest > P2WSH on regtest 1`] = `"bcrt1qxg0xyktspurnffsmtlagf80p0pjmcv5n6hgxnkgly0nvhp76ahcq825u3r"`;

exports[`bitcoin > decodes a P2WSH address on regtest > P2WSH on regtest 2`] = `"bcrt1qxg0xyktspurnffsmtlagf80p0pjmcv5n6hgxnkgly0nvhp76ahcq825u3r"`;

exports[`bitcoin > decodes a P2WSH address on regtest > P2WSH on regtest 3`] = `"bcrt1qxg0xyktspurnffsmtlagf80p0pjmcv5n6hgxnkgly0nvhp76ahcq825u3r"`;

exports[`bitcoin > decodes a P2WSH address on sisyphos > P2WSH on sisyphos 1`] = `"tb1qxg0xyktspurnffsmtlagf80p0pjmcv5n6hgxnkgly0nvhp76ahcq2n76ye"`;

exports[`bitcoin > decodes a P2WSH address on sisyphos > P2WSH on sisyphos 2`] = `"tb1qxg0xyktspurnffsmtlagf80p0pjmcv5n6hgxnkgly0nvhp76ahcq2n76ye"`;

exports[`bitcoin > decodes a P2WSH address on sisyphos > P2WSH on sisyphos 3`] = `"tb1qxg0xyktspurnffsmtlagf80p0pjmcv5n6hgxnkgly0nvhp76ahcq2n76ye"`;

exports[`bitcoin > decodes a P2WSH address on testnet > P2WSH on testnet 1`] = `"tb1qxg0xyktspurnffsmtlagf80p0pjmcv5n6hgxnkgly0nvhp76ahcq2n76ye"`;

exports[`bitcoin > decodes a P2WSH address on testnet > P2WSH on testnet 2`] = `"tb1qxg0xyktspurnffsmtlagf80p0pjmcv5n6hgxnkgly0nvhp76ahcq2n76ye"`;

exports[`bitcoin > decodes a P2WSH address on testnet > P2WSH on testnet 3`] = `"tb1qxg0xyktspurnffsmtlagf80p0pjmcv5n6hgxnkgly0nvhp76ahcq2n76ye"`;

exports[`bitcoin > decodes a Taproot address on backspin > Taproot on backspin 1`] = `"bcrt1pdz3akc5wa2gr69v3x87tfg0ka597dxqvfl6zhqx4y202y63cgw0qu6z8wq"`;

exports[`bitcoin > decodes a Taproot address on backspin > Taproot on backspin 2`] = `"bcrt1pdz3akc5wa2gr69v3x87tfg0ka597dxqvfl6zhqx4y202y63cgw0qu6z8wq"`;

exports[`bitcoin > decodes a Taproot address on backspin > Taproot on backspin 3`] = `"bcrt1pdz3akc5wa2gr69v3x87tfg0ka597dxqvfl6zhqx4y202y63cgw0qu6z8wq"`;

exports[`bitcoin > decodes a Taproot address on mainnet > Taproot on mainnet 1`] = `"bc1pdz3akc5wa2gr69v3x87tfg0ka597dxqvfl6zhqx4y202y63cgw0qxt7wp4"`;

exports[`bitcoin > decodes a Taproot address on mainnet > Taproot on mainnet 2`] = `"bc1pdz3akc5wa2gr69v3x87tfg0ka597dxqvfl6zhqx4y202y63cgw0qxt7wp4"`;

exports[`bitcoin > decodes a Taproot address on mainnet > Taproot on mainnet 3`] = `"bc1pdz3akc5wa2gr69v3x87tfg0ka597dxqvfl6zhqx4y202y63cgw0qxt7wp4"`;

exports[`bitcoin > decodes a Taproot address on perseverance > Taproot on perseverance 1`] = `"tb1pdz3akc5wa2gr69v3x87tfg0ka597dxqvfl6zhqx4y202y63cgw0q3rgpm6"`;

exports[`bitcoin > decodes a Taproot address on perseverance > Taproot on perseverance 2`] = `"tb1pdz3akc5wa2gr69v3x87tfg0ka597dxqvfl6zhqx4y202y63cgw0q3rgpm6"`;

exports[`bitcoin > decodes a Taproot address on perseverance > Taproot on perseverance 3`] = `"tb1pdz3akc5wa2gr69v3x87tfg0ka597dxqvfl6zhqx4y202y63cgw0q3rgpm6"`;

exports[`bitcoin > decodes a Taproot address on regtest > Taproot on regtest 1`] = `"bcrt1pdz3akc5wa2gr69v3x87tfg0ka597dxqvfl6zhqx4y202y63cgw0qu6z8wq"`;

exports[`bitcoin > decodes a Taproot address on regtest > Taproot on regtest 2`] = `"bcrt1pdz3akc5wa2gr69v3x87tfg0ka597dxqvfl6zhqx4y202y63cgw0qu6z8wq"`;

exports[`bitcoin > decodes a Taproot address on regtest > Taproot on regtest 3`] = `"bcrt1pdz3akc5wa2gr69v3x87tfg0ka597dxqvfl6zhqx4y202y63cgw0qu6z8wq"`;

exports[`bitcoin > decodes a Taproot address on sisyphos > Taproot on sisyphos 1`] = `"tb1pdz3akc5wa2gr69v3x87tfg0ka597dxqvfl6zhqx4y202y63cgw0q3rgpm6"`;

exports[`bitcoin > decodes a Taproot address on sisyphos > Taproot on sisyphos 2`] = `"tb1pdz3akc5wa2gr69v3x87tfg0ka597dxqvfl6zhqx4y202y63cgw0q3rgpm6"`;

exports[`bitcoin > decodes a Taproot address on sisyphos > Taproot on sisyphos 3`] = `"tb1pdz3akc5wa2gr69v3x87tfg0ka597dxqvfl6zhqx4y202y63cgw0q3rgpm6"`;

exports[`bitcoin > decodes a Taproot address on testnet > Taproot on testnet 1`] = `"tb1pdz3akc5wa2gr69v3x87tfg0ka597dxqvfl6zhqx4y202y63cgw0q3rgpm6"`;

exports[`bitcoin > decodes a Taproot address on testnet > Taproot on testnet 2`] = `"tb1pdz3akc5wa2gr69v3x87tfg0ka597dxqvfl6zhqx4y202y63cgw0q3rgpm6"`;

exports[`bitcoin > decodes a Taproot address on testnet > Taproot on testnet 3`] = `"tb1pdz3akc5wa2gr69v3x87tfg0ka597dxqvfl6zhqx4y202y63cgw0q3rgpm6"`;
